xcm/xcm-builder/src/integration_tests.rs

```diff
@@ -0,0 +1,209 @@
+						assets: vec![All],
+						dest: Parachain(other_para_id).into(),
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		let other_para_acc: AccountId = ParaId::from(other_para_id).into_account();
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - amount);
+		assert_eq!(Balances::free_balance(other_para_acc), amount);
+	});
+}
+
+/// Scenario:
+/// A parachain wants to be notified that a transfer worked correctly.
+/// It sends a `QueryHolding` after the deposit to get notified on success.
+///
+/// Asserts that the balances are updated correctly and the expected XCM is sent.
+#[test]
+fn query_holding_works() {
+	use xcm::opaque::v0::prelude::*;
+	let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+	let balances = vec![(ALICE, INITIAL_BALANCE), (para_acc.clone(), INITIAL_BALANCE)];
+	kusama_like_with_balances(balances).execute_with(|| {
+		let other_para_id = 3000;
+		let amount = 10 * ExistentialDeposit::get();
+		let query_id = 1234;
+		let weight = 4 * BaseXcmWeight::get();
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: OnlyChild.into(), // invalid destination
+					},
+					// is not triggered becasue the deposit fails
+					Order::QueryHolding {
+						query_id,
+						dest: Parachain(PARA_ID).into(),
+						assets: vec![All],
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Incomplete(weight, XcmError::FailedToTransactAsset("AccountIdConversionFailed")));
+		// there should be no query response sent for the failed deposit
+		assert_eq!(mock::sent_xcm(), vec![]);
+		assert_eq!(Balances::free_balance(para_acc.clone()), INITIAL_BALANCE - amount);
+
+		// now do a successful transfer
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: vec![All],
+						dest: Parachain(other_para_id).into(),
+					},
+					// used to get a notification in case of success
+					Order::QueryHolding {
+						query_id,
+						dest: Parachain(PARA_ID).into(),
+						assets: vec![All],
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		let other_para_acc: AccountId = ParaId::from(other_para_id).into_account();
+		assert_eq!(Balances::free_balance(other_para_acc), amount);
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - 2 * amount);
+		assert_eq!(
+			mock::sent_xcm(),
+			vec![(
+				Parachain(PARA_ID).into(),
+				Xcm::QueryResponse {
+					query_id,
+					response: Response::Assets(vec![])
+				}
+			)]
+		);
+	});
+}
+
+/// Scenario:
+/// A parachain wants to move KSM from Kusama to Statemine.
```

Comment:
Is the assumption in this test that statemine is a special parachain (sibling?)

Reply:
because specifically it is a teleport vs reserved asset transfer

Reply:
So we should also show fail for another parachain in this test?

Reply:
Perhaps this wording would make it clearer?

```suggestion
/// A parachain wants to move KSM from Kusama to Statemine, a trusted destination.
```